cadquery/assembly.py

```diff
@@ -0,0 +1,415 @@
+            
+            b = Workplane().box(1,1,1)
+            assy = Assembly(b, Location(Vector(0,0,1)), name="root")
+            
+        """
+
+        self.obj = obj
+        self.loc = loc if loc else Location()
+        self.name = name if name else str(uuid())
+        self.color = color if color else None
+        self.parent = None
+
+        self.children = []
+        self.constraints = []
+        self.objects = {self.name: self}
+
+    def _copy(self) -> "Assembly":
+        """
+        Make a deep copy of an assembly
+        """
+
+        rv = self.__class__(self.obj, self.loc, self.name, self.color)
+
+        for ch in self.children:
+            ch_copy = ch._copy()
+            ch_copy.parent = rv
+
+            rv.children.append(ch_copy)
+            rv.objects[ch_copy.name] = ch_copy
+            rv.objects.update(ch_copy.objects)
+
+        return rv
+
+    @overload
+    def add(
+        self,
+        obj: "Assembly",
+        loc: Optional[Location] = None,
+        name: Optional[str] = None,
+        color: Optional[Color] = None,
+    ) -> "Assembly":
+        """
+        add a subassembly to the current assembly.
+        
+        :param obj: subassembly to be added
+        :param loc: location of the root object (deafault: None, resulting in the location stored in the subassembly being used)
+        :param name: unique name of the root object (default: None, resulting in the name stored in the subassembly being used)
+        :param color: color of the added object (default: None, resulting in the color stored in the subassembly being used)
+        """
+        ...
+
+    @overload
+    def add(
+        self,
+        obj: AssemblyObjects,
+        loc: Optional[Location] = None,
+        name: Optional[str] = None,
+        color: Optional[Color] = None,
+    ) -> "Assembly":
+        """
+        add a subassembly to the current assembly with explicit location and name
+        
+        :param obj: object to be added as a subassembly
+        :param loc: location of the root object (deafault: None, interpreted as identity transformation)
+        :param name: unique name of the root object (default: None, resulting in an UUID being generated)
+        :param color: color of the added object (default: None)
+        """
+        ...
+
+    def add(self, arg, **kwargs):
+        """
+        add a subassembly to the current assembly.
+        """
+
+        if isinstance(arg, Assembly):
+
+            subassy = arg._copy()
+
+            subassy.loc = kwargs["loc"] if kwargs.get("loc") else arg.loc
+            subassy.name = kwargs["name"] if kwargs.get("name") else arg.name
+            subassy.color = kwargs["color"] if kwargs.get("color") else arg.color
+            subassy.parent = self
+
+            self.children.append(subassy)
+            self.objects[subassy.name] = subassy
+            self.objects.update(subassy.objects)
+
+            arg.parent = self
```

Comment:
Why is the parent of arg set here when a copy of arg was made above and already had its parent set?

Reply:
Good catch, arg should be untouched!